### PR TITLE
feat(resource): introduce resource module

### PR DESCRIPTION
**Pull request by LioRoger**
#### What type of PR is this?

type-feature
#### What this PR does / why we need it:
introduce resource module
1. make resource as a dependent module
2. introduce resource operator as resource manipulate interface
3. resource operator can be dynamic registered to adopt multi resource management needs
#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ */
+@Data
+@Entity
+@ToString
+@Table(name = "task_resource")
+public class ResourceEntity {
+    public static final String CREATE_TIME = "createTime";
+    public static final String STATUS = "status";
+    /**
+     * id for task
```
Comment (guowl3) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java:52]: Is this field supposed to represent the auto-increment ID of the resource? It doesn't seem to be related to the task

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+     * region of the resource, eg beijing/shanghai
+     */
+    @Column(name = "region", updatable = false, nullable = false)
+    private String region;
+
+    /**
+     * group of the resource, eg k8s cluster
+     */
+    @Column(name = "group_name", updatable = false, nullable = false)
+    private String groupName;
```
Comment (guowl3) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java:85]: The 'groupName' field is somewhat confusing, but I don't have a better suggestion...

File: server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceState.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * @date 2024/8/12 11:50
+ */
+public enum ResourceState {
+    // when submit create resource request, resource not available yet
+    // k8s pending/container creating / init
+    CREATING,
+    // resource is created, compute resource available
+    // task will only yield to resource with state running
+    // k8s running
+    RUNNING,
```
Comment (guowl3) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceState.java:31]: 'AVAILABLE' might be better for the resource if there are no historical compatibility costs

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java (hunk 1 above)
Comment (LioRoger) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java:52]: this filed not related to task. it's only a table unique key related to resource entity storage. even in resource module, this id is not visible to users.

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java (hunk 2 above)
Comment (LioRoger) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/resource/ResourceEntity.java:85]: region name union group name should related to a endpoint of resource operator. cloud provider is candidate name but may be too strictly. So group name is acceptable

File: server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceState.java (hunk 3 above)
Comment (LioRoger) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceState.java:31]: agreed

File: server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceEndPoint.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * endpoint of resource
+ * 
+ * @author longpeng.zlp
+ * @date 2024/8/12 11:10
+ */
+@AllArgsConstructor
+@Data
+public class ResourceEndPoint {
+    private final String resourceURL;
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceEndPoint.java:30]: resourceUrl

Comment (LioRoger) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/resource/ResourceEndPoint.java]: done

**Review by yhilmare — APPROVED** [review]
lgtm
